core/src/main/kotlin/com/alecstrong/sqlite/psi/core/psi/SqliteCompositeElementImpl.kt

```diff
@@ -26,9 +28,21 @@ internal open class SqliteCompositeElementImpl(
     return "${super.toString()}: ${(parent as SqliteCompositeElement).queryAvailable(this)}"
   }
 
+  inline fun <T> analyze(key: String, block: () -> T): T {
+    val start = System.currentTimeMillis()
+    try {
+      return block()
+    } finally {
+      val duration = System.currentTimeMillis() - start
+      analytics.put(key, analytics.getOrDefault(key, Analytics()) + duration)
```

Comment:
This isn't safe for multithread and/or concurrent use. Do all of these operations happen on a single thread?

Reply:
Yea. Which might be another thing that could speed up the compiler :D. I'll file an issue to parallelize the compiler since it's gonna require a lot of changes.

How does one even measure time taken in a concurrent world? kotlins `measure*` functions look to all also use the System clock?

Reply:
You'd either need a `ConcurrentMap<String, AtomicReference<Analytics>>` and then do a CAS loop on the `AtomicReference` or you could keep thread-local `Analytics` and aggregate them together into a single value when requested.

Reply:
Oh unless you mean how do you measure work that's happening in parallel. You'd probably want to aggregate the wall clock time of the operation (start to finish) and the computation time (amount of time spent on each thread). The more threads/CPUs that you have the lower the wall clock time would go, but the computation time should remain consistent based on the size of the work.

Reply:
Makes sense - is there an api to get computation time for some block that isn't effected by preemption?

Reply:
Doubtful!

Reply:
Some stuff on https://docs.oracle.com/javase/7/docs/api/java/lang/management/ThreadMXBean.html